v1/src/main/java/com/google/cloud/teleport/bigtable/BeamRowToBigtableFn.java

```diff
@@ -148,41 +180,116 @@ public void processElement(
+      if (this.splitLargeRows && cellsProcessed % maxMutationsPerRequest == 0) {
+        // Send a MutateRow request when we have accumulated max mutations per request.
+        out.output(KV.of(rowkey, mutations.build()));
+        mutations = ImmutableList.builder();
       }
     }
-
     // Flush any remaining mutations.
     ImmutableList remainingMutations = mutations.build();
     if (!remainingMutations.isEmpty()) {
       out.output(KV.of(rowkey, remainingMutations));
     }
   }
 
+  private HashMap<String, Long> getColumnWritetimes(Row row, List<Field> nonKeyColumns) {
+    // Writetime values will be joined to the other mutations as timestamps.
+    HashMap<String, Long> columnTimestamps = new HashMap<>();
+    Pattern writetimeColumn = Pattern.compile(String.format("%s\\((.*)\\)", WRITETIME_COLUMN));
+    for (Field field : nonKeyColumns) {
+      if (field.getName().contains(WRITETIME_COLUMN)) {
+        // Get column name of writetime column, e.g. "writetime(column1)" will be the writetime for
+        // "column1".
+        Matcher matcher = writetimeColumn.matcher(field.getName());
+        matcher.matches();
+        String columnName = matcher.group(1);
+
+        // Get timestamp and normalize as necessary.
+        Long rawTimestamp = (Long) row.getValue(field.getName());
+        Long normalizedTimestamp = normalizeToBigtableTimestamp(rawTimestamp);
+
+        columnTimestamps.put(columnName, normalizedTimestamp);
+      }
+    }
+    return columnTimestamps;
+  }
+
+  /**
+   * Bigtable requires timestamps to be a microsecond value in millisecond precision. Cassandra
+   * timestamps can either be microsecond or millisecond precision. Therefore, this function does
+   * two things: 1. casts microsecond timestamps to millisecond precision by setting last 3 digits
+   * to 0. 2. casts millisecond timestamps to microsecond value by padding 3 more 0s to the end.
+   *
+   * @param timestamp in either micro or millisecond format
+   * @return microsecond value timestamp in millisecond precision
+   */
+  Long normalizeToBigtableTimestamp(Long timestamp) {
+    if (timestamp < 0L) {
+      throw new IllegalArgumentException("Timestamp " + timestamp + " is smaller than 0.");
+    }
+    Long newTimestamp = timestamp;
+    Long nowInMillis = Instant.now().toEpochMilli();
+    Long nowInMicros = nowInMillis * 1000;
+
+    if (timestamp <= nowInMillis) {
+      // Timestamp is in millisecond format. If so, pad 3 more 0s to the end.
+      newTimestamp *= 1000;
+    } else if (timestamp <= nowInMicros) {
+      // Time is in microsecond format. If so, set last 3 digits to 0.
+      newTimestamp /= 1000;
+      newTimestamp *= 1000;
+    } else {
+      // In this case, the timestamp is set in the future. It is not recommended to do this due to
+      // implications around Bigtable garbage collection. However, there is nothing wrong with this
+      // technically so this action is not blocked.
+      LOG.warn(
```

Comment:
The case here now is that this time is > 100 years into the future in microsecond timestamps. At this point, I think it'll be more important to call out this behavior than to modify it since we're not sure what the intention is anymore.